### ISSUE TITLE
aggregate queries should respect user-provided page-size

### ISSUE THREAD

**Issue by denesb** [scylladb/scylladb#20612]
From @fee-mendes:

For quite sometime now, whenever an aggregation query gets executed, we no longer respect the client-side paging settings. We probably never realized it until now, because we hardcode `DEFAULT_COUNT_PAGE_SIZE = 10000;`.

In https://github.com/scylladb/scylladb/commit/f7aa2a58bd74eb61ed247ffbfcb61b6d08175a9f (hello 2015!) we introduced the following logic to `select_statement::execute`:

```cpp
    // An aggregation query will never be paged for the user, but we always page it internally to avoid OOM.
    // If we user provided a page_size we'll use that to page internally (because why not), otherwise we use our default
    // Note that if there are some nodes in the cluster with a version less than 2.0, we can't use paging (CASSANDRA-6707).
    if (_selection->is_aggregate() && page_size <= 0) {
        page_size = DEFAULT_COUNT_PAGE_SIZE;
    }
```

Notably, the description matches the condition intent: When we receive an unpaged aggregation, we must internally page it.

Things changed in https://github.com/scylladb/scylladb/commit/acf4eadf88ccd0bebfa22d05d162f82485ed6491 (woot 2018!). We inadvertently changed the condition to:

```cpp
    const bool aggregate = _selection->is_aggregate();
    const bool nonpaged_filtering = restrictions_need_filtering && page_size <= 0;
    if (aggregate || nonpaged_filtering) {
        page_size = DEFAULT_COUNT_PAGE_SIZE;
    }
```

With this change, we always page filtering queries received as unpaged by the client. This is good, as it avoids OOMs. **However**, this change effectively invalidated the original comment from 2015: **We no longer rely on the user supplied `page_size`, and all and any aggregates will internally page following `DEFAULT_COUNT_PAGE_SIZE`.**

Going even further, in 5.2 we introduced https://github.com/scylladb/scylladb/pull/11053. However, the original behavior of aggregation queries still stands: `An aggregation query will never be paged for the user (...)`.

With Empty Replica Pages we internally emit an empty page as we scan through long tombstone prefixes, leading to the following observations:

- If we never propagate empty pages to clients, then why add empty pages overhead in the first place? Sure, the replica won't timeout... But the client will wait longer and increase the chances of a coordinator timeout.
- If the intention is to propagate these pages to clients (and thus save the aggregation from timing out), then why we are not doing that?

Moreover, note that all this may end up creating weird and different Paging behaviors for aggregates. For example:

1. A wide partition with many live rows will internally page at every 10K rows irrespective of the client-side `page_size`
2. A wide partition with long tombstone prefixes will internally page at every `query_tombstone_page_limit` (defaults to 10K)
3. A wide partition with many live rows and tombstones will internally page a LOT as we now have both internal paging and empty pages 

Although paging is most of the times beneficial, paging too much may actually be DETRIMENTAL to performance. This should very easily be observed with - for example - a `query_tombstone_page_limit: 1000` and running a `COUNT(*)` on a 1M tombstone prefix.

The problem is that our defaults may sometime be too strict: For either of the 3 above cases, depending on the number of rows in a partition, we can easily end up paging too much considerably increasing our coordinator latencies.

----

Summary: we used to respect the client's page size for the internal paging done for aggregates. This was broken and now we always use `DEFAULT_COUNT_PAGE_SIZE`. I think we should restore the previous behavior, so the client (user) has a say in the page size, which has an effect on the performance characteristics of the aggregate query.

**Comment by ptrsmrn** [comment]
> It should've been fixed by [08f3219](https://github.com/scylladb/scylladb/commit/08f3219cb8c4bd781454a46ee5adf0203cb7d6fc)

Do we have any test for this scenario? Can you link it here? If not, can you write one?

**Comment by denesb** [comment]
> It should've been fixed by [08f3219](https://github.com/scylladb/scylladb/commit/08f3219cb8c4bd781454a46ee5adf0203cb7d6fc)

Not fully: `std::min(page_size, internal_paging_size);` it still enforces an upper limit on the page-size. Larger page-size is exactly what you might want to do if you want to improve the performance of the query. Note that we have an internal hard-limit of 1MB size anyway, so you don't have to worry about the user providing too large page-size and the system running out of memory.

**Comment by denesb** [comment]
> okay, would you be satisfied by using `page_size` (user defined) if it's a positive integer, or `internal_paging_size` otherwise?

Yes, that sounds good.
